rust/actyx/ax-aql/src/language/type_check/drill_down_type.rs

```diff
@@ -0,0 +1,443 @@
+                        .map(|err_msg| format!("- {}", err_msg))
+                        .collect::<Vec<_>>()
+                        .join("\n");
+                    return Err(error_messages.trim().into());
+                }
+
+                // NOTE: Using Option to emulate bottom type; consider introducing bottom type
+                let union = type_results
+                    .into_iter()
+                    .filter_map(|x| match x {
+                        Ok(x) => Some(x),
+                        Err(_) => None,
+                    })
+                    .fold(None, |acc, item| match acc {
+                        None => Some(item),
+                        Some(prev) => Some(Type::Union(Arc::new((prev, item)))),
+                    });
+
+                let union = union
+                    .expect("the type_results to be non-empty because it is derived from a non-empty cases")
+                    .collapse()
+                    .map_err(|e| e.to_string())?;
+
+                return Ok(union);
+            }
+        };
+
+        drill_down(current_type, [&[new_index], rest].concat().as_slice())
+    }
+
+    // start drilling_down
+    let Ind { head: _, tail, .. } = ind;
+    let result = drill_down(current_type, tail);
+
+    match result {
+        Ok(x) => Ok(x.collapse().map_err(|e| e.to_string())?),
+        Err(x) => {
+            if tail.len() > 1 {
+                // create a root-level error message
+                let root_lv_err = format!("{:?} cannot be accessed by {:?}", current_type, ind);
+                // append with indented deeper_lv_err
+                let deeper_lv_err = x
+                    .split('\n')
+                    .map(|line| format!("  {:?}", line))
+                    .collect::<Vec<_>>()
+                    .join("\n");
+
+                Err([root_lv_err, deeper_lv_err].join("\n"))
+            } else {
+                Err(x)
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::sync::Arc;
+
+    use super::*;
+    use crate::{TypeAtom, Var};
+
+    #[test]
+    fn array_by_number() {
+        assert_eq!(
+            drill_down_type_by_index(
+                &Type::Array(Arc::new(Type::Atom(TypeAtom::Number(None)))),
+                &Ind {
+                    head: SimpleExpr::Variable(Var("_".to_string())).into(),
+                    tail: vec![Index::Number(0_u64)].try_into().unwrap(),
+                },
+            ),
+            Ok(Type::Union(Arc::new((
+                Type::Atom(TypeAtom::Number(None)),
+                Type::NoValue
+            ))))
+        );
+
+        assert!(drill_down_type_by_index(
+            &Type::Array(Arc::new(Type::Atom(TypeAtom::Number(None)))),
+            &Ind {
+                head: SimpleExpr::Variable(Var("_".to_string())).into(),
+                tail: vec![Index::String(String::from("some_string"))].try_into().unwrap(),
+            },
+        )
+        .is_err());
```

Comment:
I'd add the second parameter to `assert!` with a short explanation of what failed